File: modules/core/app/model/matching/Match.java
package model.matching;

public abstract class Match<T> {
  
  protected T userArg;
  protected T sampleArg;
  
  protected MatchType matchType = MatchType.FAILURE;
  
  public Match(T theUserArg, T theSampleArg) {
    userArg = theUserArg;
    sampleArg = theSampleArg;
    
    if(theUserArg == null && theSampleArg != null)
      matchType = MatchType.ONLY_SAMPLE;
    else if(theUserArg != null && theSampleArg == null)
      matchType = MatchType.ONLY_USER;
    else
      matchType = analyze(theUserArg, theSampleArg);
  }
  
  public String getBSClass() {
    return matchType == MatchType.SUCCESSFUL_MATCH ? "success" : "warning";
  }
  
  public String getExplanation() {
    switch(matchType) {
    case FAILURE:
      return "Es ist ein Fehler aufgetreten.";
    case ONLY_USER:
      return "Angegeben, aber nicht in der Musterlösung vorhanden!";
    case ONLY_SAMPLE:
      return "Nur in der Musterlösung, nicht in der Lernerlösung vorhanden!";
    case UNSUCCESSFUL_MATCH:
      return "Fehler beim Abgleich.";
    case SUCCESSFUL_MATCH:
      return "Korrekt.";
    default:
      throw new IllegalArgumentException();
    }
  }
  
  public MatchType getMatchType() {
    return matchType;
  }
  
  public T getSampleArg() {
    return sampleArg;
  }
  
  public T getUserArg() {
    return userArg;
  }
  
  public boolean isSuccessful() {
    return matchType == MatchType.SUCCESSFUL_MATCH;
  }
  
  protected abstract MatchType analyze(T theArg1, T theArg2);
  
}

File: modules/core/app/model/matching/Match.scala
package model.matching

import play.twirl.api.Html

class Match[T](val userArg: Option[T], val sampleArg: Option[T], val size: Int) {

  val matchType: MatchType = (userArg, sampleArg) match {
    case (None, None) => throw new IllegalArgumentException("At least one arg of a match must not be None!")
    case (None, Some(_)) => MatchType.ONLY_SAMPLE
    case (Some(_), None) => MatchType.ONLY_USER
    case (Some(_), Some(_)) => analyze(userArg.get, sampleArg.get)
  }

  def getBSClass: String = if (matchType == MatchType.SUCCESSFUL_MATCH) "success" else "warning"

  def explanation = List(matchType match {
    case MatchType.FAILURE => "Es ist ein Fehler aufgetreten."
    case MatchType.ONLY_USER => "Angabe ist falsch!"
    case MatchType.ONLY_SAMPLE => "Angabe fehlt!"
    case MatchType.UNSUCCESSFUL_MATCH => "Fehler beim Abgleich."
    case MatchType.SUCCESSFUL_MATCH => "Korrekt."
  })

  lazy val isSuccessful: Boolean = matchType == MatchType.SUCCESSFUL_MATCH

  def analyze(arg1: T, arg2: T): MatchType = MatchType.SUCCESSFUL_MATCH

  def describeArg(arg: T): Html = new Html(
    s"""
<td>
  <span class="text-${if (isSuccessful) "success" else "danger"}>${arg.toString}</span>
</td>""")

  private def describeArgOption(arg: Option[T]) = arg match {
    case Some(a) => describeArg(a)
    case None => new Html("<td>&nbsp;</td>" * size)
  }

  def describeUserArg: Html = describeArgOption(userArg)

  def describeSampleArg: Html = describeArgOption(sampleArg)

}

